benchmark/tools/compare.py

```diff
@@ -0,0 +1,261 @@
+            result[key] = baseline[key]
+    return result
+
+
+def extract_benchmark_results(
+    input: dict, benchmarks: dict, case_key: tuple, context: str | None
+):
+    for key, value in input.items():
+        benchmark_name = key if context is None else "{}.{}".format(context, key)
+        if key in map(lambda x: x + suffix, comparison_keys):
+            benchmark_name = benchmark_name[: -len(suffix)]
+            if benchmark_name not in benchmarks.keys():
+                benchmarks[benchmark_name] = []
+            benchmarks[benchmark_name].append((case_key, value))
+        elif isinstance(value, dict):
+            extract_benchmark_results(value, benchmarks, case_key, benchmark_name)
+
+
+def is_outlier(value: float):
+    return math.fabs(math.log(value)) > math.log(1.0 + args.outlier_threshold / 100)
+
+
+baseline_json = parse_json_matrix(args.baseline)
+comparison_json = parse_json_matrix(args.comparison)
+warn_on_inconsistent_keys(baseline_json, comparison_json, "root")
+
+results = {}
+
+for key in baseline_json.keys():
+    results[key] = compare(baseline_json[key], comparison_json[key], key_to_str(key))
+
+outliers = {}
+if args.outliers:
+    benchmarks = {}
+    for key, value in results.items():
+        extract_benchmark_results(value, benchmarks, key, None)
+    for benchmark_name, benchmark_results in benchmarks.items():
+        outlier = sorted(
+            [
+                (case_key, value)
+                for case_key, value in benchmark_results
+                if is_outlier(value)
+            ],
+            key=lambda x: math.fabs(math.log(x[1])),
+            reverse=True,
+        )
+        outliers[benchmark_name] = outlier[: min(len(outlier), args.outlier_count)]
+
+if args.output == "json":
+    print(
+        json.dumps(
+            {
+                "results": [value for _, value in results.items()],
+                "outliers": {
+                    key: [
+                        {"value": ratio_value} + key_to_json(case_key)
+                        for (case_key, ratio_value) in value
+                    ]
+                    for key, value in outliers.items()
+                    if len(value) > 0
+                },
+            },
+            indent=4,
+        )
+    )
+else:
+    columns = ["benchmark", "testcase", "ratio"]
+    only_first = args.output == "markdown"
+    table = pd.DataFrame(
+        sum(
+            [
+                [
+                    (
+                        key if i == 0 or not only_first else "",
+                        key_to_str(value[0]),
+                        value[1],
+                    )
+                    for i, value in enumerate(values)
+                ]
+                for key, values in benchmarks.items()
+            ],
+            [],
+        ),
+        columns=columns,
+    )
+    if args.output == "csv":
+        table.to_csv(sys.stdout, index=False)
+    else:
+        table.to_markdown(sys.stdout, index=False)
```

Comment:
These two have empty output if `outliers` is not set. That is inconsistent with the json output.